src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/Common/DbConnectionOptions.cs

```diff
@@ -189,60 +34,61 @@ internal bool HasBlankPassword
+                        return (_parsetable.ContainsKey(KEY.User_ID) && !ADP.IsEmpty(_parsetable[KEY.User_ID])) || 
+                               (_parsetable.ContainsKey(SYNONYM.UID) && !ADP.IsEmpty(_parsetable[SYNONYM.UID]));
                     }
                 }
                 return false;
             }
         }
 
-        internal bool HasPersistablePassword
+        internal string ExpandKeyword(string keyword, string replacementValue)
         {
-            get
+            // preserve duplicates, updated keyword value with replacement value
+            // if keyword not specified, append to end of the string
+            bool expanded = false;
+            int copyPosition = 0;
+
+            StringBuilder builder = new(_usersConnectionString.Length);
+            for (NameValuePair current = KeyChain; null != current; current = current.Next)
             {
-                if (HasPasswordKeyword)
+                if ((current.Name == keyword) && (current.Value == this[keyword]))
                 {
-                    return ConvertValueToBoolean(KEY.Persist_Security_Info, false);
+                    // only replace the parse end-result value instead of all values
+                    // so that when duplicate-keywords occur other original values remain in place
+                    AppendKeyValuePairBuilder(builder, current.Name, replacementValue);
+                    builder.Append(';');
+                    expanded = true;
+                }
+                else
+                {
+                    builder.Append(_usersConnectionString, copyPosition, current.Length);
                 }
-                return true; // no password means persistable password so we don't have to munge
+                copyPosition += current.Length;
             }
-        }
-
-        public bool IsEmpty
-        {
-            get { return (null == KeyChain); }
-        }
-
-        internal Hashtable Parsetable
-        {
-            get { return _parsetable; }
-        }
-
-        public ICollection Keys
-        {
-            get { return _parsetable.Keys; }
-        }
 
-        public string this[string keyword]
-        {
-            get { return (string)_parsetable[keyword]; }
+            if (!expanded)
+            {
+                AppendKeyValuePairBuilder(builder, keyword, replacementValue);
+            }
+            return builder.ToString();
         }
 
-        internal static void AppendKeyValuePairBuilder(StringBuilder builder, string keyName, string keyValue, bool useOdbcRules)
+        internal static void AppendKeyValuePairBuilder(StringBuilder builder, string keyName, string keyValue, bool useOdbcRules = false)
         {
             ADP.CheckArgumentNull(builder, "builder");
             ADP.CheckArgumentLength(keyName, "keyName");
```

Comment:
not your changes, but can we define constant strings and refer them here please?

Reply:
Improved using `nameof()` function.